File: androidApp/src/main/kotlin/com/malaria/android/ui/history/components/SessionRowView.kt
package com.malaria.android.ui.history.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

/**
 * One row in the Sessions list. Also embedded in PredictionDetailView's
 * "Session" section as a navigable affordance.
 *
 * Mirrors `iosApp/Views/History/Components/SessionRowView.swift`.
 */
@Composable
fun SessionRowView(
    stats: SessionStats,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = 4.dp, bottom = 4.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = stats.displayLabel,
                style = MaterialTheme.typography.titleMedium,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f),
            )
            Text(
                text = "${stats.count} cells",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Row(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.Top,
        ) {
            StatBadge(label = "Parasitized", value = "${stats.parasitizedCount}")
            StatBadge(label = "Gray zone", value = "${stats.grayZoneCount}")
            StatBadge(label = "Mean p", value = stats.meanParasitizedFormatted)
        }
        Text(
            text = stats.dateRangeLabel,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
private fun StatBadge(label: String, value: String) {
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelMedium,
        )
    }
}

File: iosApp/Views/History/Components/SessionRowView.swift
import SwiftUI
@preconcurrency import Shared

/// One row in the Sessions list. Also embedded in PredictionDetailView's
/// "Session" section as a navigable affordance.
struct SessionRowView: View {
    let stats: SessionStats

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(alignment: .firstTextBaseline) {
                Text(stats.displayLabel)
                    .font(.headline)
                    .lineLimit(1)
                Spacer()
                Text("\(stats.count) cells")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
            HStack(spacing: 12) {
                statBadge(label: "Parasitized", value: "\(stats.parasitizedCount)")
                statBadge(label: "Gray zone", value: "\(stats.grayZoneCount)")
                statBadge(label: "Mean p", value: stats.meanParasitizedFormatted)
            }
            Text(stats.dateRangeLabel)
                .font(.caption2)
                .foregroundStyle(.secondary)
        }
        .padding(.vertical, 4)
    }

    private func statBadge(label: String, value: String) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
            Text(value)
                .font(.caption.weight(.semibold))
        }
    }
}

/// Aggregated stats for one session, computed in Swift from a list of
/// predictions. Session-stats live in shared Kotlin in v1.1 (spec §13)
/// but we keep them client-side for Phase 10 — no read-path mutation
/// needed in repos.
struct SessionStats: Identifiable, Hashable {
    let sessionId: String
    let sessionLabel: String?
    let count: Int
    let parasitizedCount: Int
    let grayZoneCount: Int
    let meanParasitizedProb: Double
    let earliest: Date
    let latest: Date

    var id: String { sessionId }

    var displayLabel: String {
        if let label = sessionLabel, !label.isEmpty {
            return label
        }
        return "Session \(String(sessionId.prefix(8)))"
    }

    var meanParasitizedFormatted: String {
        "\(Int((meanParasitizedProb * 100).rounded()))%"
    }

    var dateRangeLabel: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .short
        if Calendar.current.isDate(earliest, inSameDayAs: latest) {
            let dayFmt = DateFormatter()
            dayFmt.dateStyle = .medium
            let timeFmt = DateFormatter()
            timeFmt.timeStyle = .short
            return "\(dayFmt.string(from: earliest)) · \(timeFmt.string(from: earliest))–\(timeFmt.string(from: latest))"
        } else {
            return "\(formatter.string(from: earliest)) – \(formatter.string(from: latest))"
        }
    }

    /// Compute stats from an arbitrary list of predictions in one session.
    /// Uses `Threshold.shared.GRAY_ZONE_*` for gray-zone classification so
    /// the count matches the rest of the app exactly.
    static func from(predictions: [Prediction]) -> SessionStats? {
        guard let first = predictions.first else { return nil }
        let sortedByTime = predictions.sorted { $0.timestamp < $1.timestamp }
        let count = predictions.count
        let parasitizedCount = predictions.filter { $0.label == "Parasitized" }.count
        let low = Double(Threshold.shared.GRAY_ZONE_LOW)
        let high = Double(Threshold.shared.GRAY_ZONE_HIGH)
        let grayZoneCount = predictions.filter { $0.parasitizedProb >= low && $0.parasitizedProb <= high }.count
        let meanProb = predictions.map(\.parasitizedProb).reduce(0, +) / Double(count)
        return SessionStats(
            sessionId: first.sessionId,
            sessionLabel: first.sessionLabel,
            count: count,
            parasitizedCount: parasitizedCount,
            grayZoneCount: grayZoneCount,
            meanParasitizedProb: meanProb,
            earliest: sortedByTime.first?.timestamp ?? first.timestamp,
            latest: sortedByTime.last?.timestamp ?? first.timestamp
        )
    }

    /// Group a flat prediction list into per-session stats, newest session
    /// first (latest timestamp).
    static func grouped(_ predictions: [Prediction]) -> [SessionStats] {
        let groups = Dictionary(grouping: predictions, by: \.sessionId)
        return groups.compactMap { _, list in SessionStats.from(predictions: list) }
            .sorted { $0.latest > $1.latest }
    }
}
